src/interface/update.hpp

```diff
@@ -13,44 +13,94 @@
-Real EstimateTimestep(std::shared_ptr<MeshBlockData<Real>> &rc);
+template <typename T>
+TaskStatus FluxDivergence(std::shared_ptr<T> &in, std::shared_ptr<T> &dudt_obj);
 
-} // namespace Update
+template <typename F, typename T>
+void UpdateData(const std::vector<F> &flags, T &in, T &dudt, const Real dt, T &out) {
+  const auto &in_pack = in->PackVariables(flags);
+  const auto &out_pack = out->PackVariables(flags);
+  const auto &dudt_pack = dudt->PackVariables(flags);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "UpdateMeshData", DevExecSpace(), 0, in_pack.GetDim(5) - 1, 0,
+      in_pack.GetDim(4) - 1, 0, in_pack.GetDim(3) - 1, 0, in_pack.GetDim(2) - 1, 0,
+      in_pack.GetDim(1) - 1,
+      KOKKOS_LAMBDA(const int b, const int l, const int k, const int j, const int i) {
+        out_pack(b, l, k, j, i) = in_pack(b, l, k, j, i) + dt * dudt_pack(b, l, k, j, i);
+      });
+}
+
+template <typename T>
+void UpdateIndependentData(T &in, T &dudt, const Real dt, T &out) {
+  UpdateData(std::vector<MetadataFlag>({Metadata::Independent}), in, dudt, dt, out);
+}
+
+template <typename F, typename T>
+void AverageData(const std::vector<F> &flags, T &c1, T &c2, const Real wgt1) {
+  const auto &c1_pack = c1->PackVariables(flags);
+  const auto &c2_pack = c2->PackVariables(flags);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "AverageMeshData", DevExecSpace(), 0, c1_pack.GetDim(5) - 1,
+      0, c1_pack.GetDim(4) - 1, 0, c1_pack.GetDim(3) - 1, 0, c1_pack.GetDim(2) - 1, 0,
+      c1_pack.GetDim(1) - 1,
+      KOKKOS_LAMBDA(const int b, const int l, const int k, const int j, const int i) {
+        c1_pack(b, l, k, j, i) =
+            wgt1 * c1_pack(b, l, k, j, i) + (1 - wgt1) * c2_pack(b, l, k, j, i);
+      });
+}
 
-namespace FillDerivedVariables {
+template <typename T>
+void AverageIndependentData(T &c1, T &c2, const Real wgt1) {
+  AverageData(std::vector<MetadataFlag>({Metadata::Independent}), c1, c2, wgt1);
+}
 
-using FillDerivedFunc = void(std::shared_ptr<MeshBlockData<Real>> &);
-void SetFillDerivedFunctions(FillDerivedFunc *pre, FillDerivedFunc *post);
-TaskStatus FillDerived(std::shared_ptr<MeshBlockData<Real>> &rc);
+template <typename T>
+TaskStatus EstimateTimestep(std::shared_ptr<T> &rc) {
+  Real dt_min = std::numeric_limits<Real>::max();
+  for (const auto &pkg : rc->GetGridPointer()->packages) {
+    Real dt = pkg.second->EstimateTimestep(rc);
+    dt_min = std::min(dt_min, dt);
+  }
+  rc->SetAllowedDt(dt_min);
+  return TaskStatus::complete;
+}
 
-} // namespace FillDerivedVariables
+template <typename T>
+TaskStatus FillDerived(std::shared_ptr<T> &rc) {
+  auto pm = rc->GetGridPointer();
+  for (const auto &pkg : pm->packages) {
+    pkg.second->PreFillDerived(rc);
+  }
+  for (const auto &pkg : pm->packages) {
+    pkg.second->FillDerived(rc);
+  }
+  for (const auto &pkg : pm->packages) {
+    pkg.second->PostFillDerived(rc);
+  }
+  return TaskStatus::complete;
+}
+
```

Comment:
From a conceptual point of view, is my understanding correct that the current implementation means that a downstream app either enrolls either `MeshData` or `MeshBlockData` functions and determine which one to use by forwarding the appropriate container in the driver?
It'd be great if we'd have one more paragraph in the doc highlighting this "unify" abstraction and what this mean for enrolled functions.

Reply:
In fact, a downstream app can make use of _both_ forms, so long as the appropriate tasks are called.

I'll add a bit to the docs.